<a href="https://colab.research.google.com/github/atoothman/data71200/blob/main/DATA_71200_Project_1_Mushroom_Classification_Toothman.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DATA 71200 - Project 1 Mushroom Classification

The Mushroom dataset used was found on the UC Irvine Machine Learning Repository [https://archive.ics.uci.edu/dataset/73/mushroom](https://archive.ics.uci.edu/dataset/73/mushroom)

The dataset includes 23 species of gilled mushrooms in the Agaricus and Lepiota family that were classificed as edible or poisonous. Samples identifed as unknown or not recommended were combined into poisonous.

The dataset is comprised of 22 features. Two features are binary and 20 are categorical.

The aim of this model is to predict if the mushroom is poisonous or edible based on its physical characteristics. The target is categorical.

# Step 1: Load Dataset

In [ ]:
import pandas as pd

# URL of the mushroom dataset from UC Irving ML repository
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.data"



# Load the dataset into a DataFrame
mushroom_df = pd.read_csv(url, header=None, index_col=False, names=[
    'poisonous', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
    'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color', 'stalk-shape',
    'stalk-root', 'stalk-surface-above-ring', 'stalk-surface-below-ring',
    'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color',
    'ring-number', 'ring-type', 'spore-print-color', 'population', 'habitat'])

# Display the DataFrame
display(mushroom_df)


poisonous cap-shape cap-surface cap-color bruises odor gill-attachment  \
0            p         x           s         n       t    p               f   
1            e         x           s         y       t    a               f   
2            e         b           s         w       t    l               f   
3            p         x           y         w       t    p               f   
4            e         x           s         g       f    n               f   
...        ...       ...         ...       ...     ...  ...             ...   
8119         e         k           s         n       f    n               a   
8120         e         x           s         n       f    n               a   
8121         e         f           s         n       f    n               a   
8122         p         k           y         n       f    y               f   
8123         e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

In [ ]:
# Show dataset info
mushroom_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

# Check for missing values

On the UCI ML page for mushrooms it says there are missing values in "stalk-root". Next steps are searching for missing values.

In [ ]:
# Check for missing values in mushroom_df
missing_values = mushroom_df.isnull().sum()

# Print missing values in mushroom_df
print("Missing Values:")
print(missing_values)

Missing Values:
class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64


Code is showing there are no missing values for "stalk-root" under NaN

Examining the UCI ML page further shows that "stalk-root" has the following:
bulbous=b,club=c,cup=u,equal=e, rhizomorphs=z,rooted=r,**missing=?**

"?" is encoded to represent the missing values.

Checking for occurrences of "?" in each column

In [ ]:
# Check missing values encoded as '?'
missing_values = mushroom_df.isin(['?']).sum()

# Print missing values
print("Missing values:")
print(missing_values)

Missing values:
class                          0
cap-shape                      0
cap-surface                    0
cap-color                      0
bruises                        0
odor                           0
gill-attachment                0
gill-spacing                   0
gill-size                      0
gill-color                     0
stalk-shape                    0
stalk-root                  2480
stalk-surface-above-ring       0
stalk-surface-below-ring       0
stalk-color-above-ring         0
stalk-color-below-ring         0
veil-type                      0
veil-color                     0
ring-number                    0
ring-type                      0
spore-print-color              0
population                     0
habitat                        0
dtype: int64


Stalk-root has 2480 missing values.

Before proceeding to splitting the data into test and training sets, the missing values should be addressed.

It was elected to remove the column "stalk-root" as removing 2480 rows might have a larger impact on the model then removing this feature as there are a significant amount of other features to work with.

In [ ]:
# Remove 'stalk-root' from mushroom_df and create modified dataframe = mushroom_df_v2
mushroom_df_v2 = mushroom_df.drop(['stalk-root'],axis=1)

display(mushroom_df_v2)

poisonous cap-shape cap-surface cap-color bruises odor gill-attachment  \
0            p         x           s         n       t    p               f   
1            e         x           s         y       t    a               f   
2            e         b           s         w       t    l               f   
3            p         x           y         w       t    p               f   
4            e         x           s         g       f    n               f   
...        ...       ...         ...       ...     ...  ...             ...   
8119         e         k           s         n       f    n               a   
8120         e         x           s         n       f    n               a   
8121         e         f           s         n       f    n               a   
8122         p         k           y         n       f    y               f   
8123         e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 22 columns]

# One-Hot Encoding of categorical data

The mushroom data set is comprised of 19 categorical features

In [93]:
# List of categorical columns omiting "stalk-root"
categorical_columns = ['poisonous', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor','gill-attachment', 'gill-spacing', 'gill-size', 'gill-color', 'stalk-shape', 'stalk-surface-above-ring', 'stalk-surface-below-ring', 'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number', 'ring-type', 'spore-print-color', 'population', 'habitat']

# One-hot encode categorical columns
mushroom_df_encoded = pd.get_dummies(mushroom_df_v2, columns=categorical_columns, drop_first=True)

# Display the encoded DataFrame
display(mushroom_df_encoded)


poisonous_p  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0            True        False        False        False        False   
1           False        False        False        False        False   
2           False        False        False        False        False   
3            True        False        False        False        False   
4           False        False        False        False        False   
...           ...          ...          ...          ...          ...   
8119        False        False        False         True        False   
8120        False        False        False        False        False   
8121        False        False         True        False        False   
8122         True        False        False         True        False   
8123        False        False        False        False        False   

      cap-shape_x  cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  \
0            True          False           True          False        False   
1            True          False           True          False        False   
2           False          False           True          False        False   
3            True          False          False           True        False   
4            True          False           True          False        False   
...           ...            ...            ...            ...          ...   
8119        False          False           True          False        False   
8120         True          False           True          False        False   
8121        False          False           True          False        False   
8122        False          False          False           True        False   
8123         True          False           True          False        False   

      ...  population_n  population_s  population_v  population_y  habitat_g  \
0     ...         False          True         False         False      False   
1     ...          True         False         False         False       True   
2     ...          True         False         False         False      False   
3     ...         False          True         False         False      False   
4     ...         False         False         False         False       True   
...   ...           ...           ...           ...           ...        ...   
8119  ...         False         False         False         False      False   
8120  ...         False         False          True         False      False   
8121  ...         False         False         False         False      False   
8122  ...         False         False          True         False      False   
8123  ...         False         False         False         False      False   

      habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0         False      False      False       True      False  
1         False      False      False      False      False  
2         False       True      False      False      False  
3         False      False      False       True      False  
4         False      False      False      False      False  
...         ...        ...        ...        ...        ...  
8119       True      False      False      False      False  
8120       True      False      False      False      False  
8121       True      False      False      False      False  
8122       True      False      False      False      False  
8123       True      False      False      False      False  

[8124 rows x 92 columns]

# Step 2: Divide into a training set and testing set



In [104]:
# Utilize scikitlearn for a stratify split
from sklearn.model_selection import StratifiedShuffleSplit

# Separate the features (X) and the target variable (y) aka poisonous_p
X = mushroom_df_encoded.drop('poisonous_p', axis=1)
y = mushroom_df_encoded['poisonous_p']

# Create StratifiedShuffleSplit with 1 split, 20% test size, 80% training and random state 42
split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
# Split data
for train_index, test_index in split.split(X, y):
    X_train = X.iloc[train_index]
    X_test = X.iloc[test_index]
    y_train = y.iloc[train_index]
    y_test = y.iloc[test_index]

# Check to confirm successful 80 : 20 split
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (6499, 91)
X_test: (1625, 91)
y_train: (6499,)
y_test: (1625,)


Confirming the split was successful. The training set has 80% of the original 8124 instances, while the testing set has 20% of the original instances. 91 in the X-train and X_test show the number of columns/features after one-hot encoding.



# Step 3: Explore the training set



In [105]:
# Training set infomation on X_train
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6499 entries, 5249 to 2411
Data columns (total 91 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   cap-shape_c                 6499 non-null   bool 
 1   cap-shape_f                 6499 non-null   bool 
 2   cap-shape_k                 6499 non-null   bool 
 3   cap-shape_s                 6499 non-null   bool 
 4   cap-shape_x                 6499 non-null   bool 
 5   cap-surface_g               6499 non-null   bool 
 6   cap-surface_s               6499 non-null   bool 
 7   cap-surface_y               6499 non-null   bool 
 8   cap-color_c                 6499 non-null   bool 
 9   cap-color_e                 6499 non-null   bool 
 10  cap-color_g                 6499 non-null   bool 
 11  cap-color_n                 6499 non-null   bool 
 12  cap-color_p                 6499 non-null   bool 
 13  cap-color_r                 6499 non-null   bool 
 14  cap-color_

In [108]:
# Training set description of X_train
X_train.describe()

cap-shape_c cap-shape_f cap-shape_k cap-shape_s cap-shape_x  \
count         6499        6499        6499        6499        6499   
unique           2           2           2           2           2   
top          False       False       False       False       False   
freq          6495        3982        5846        6472        3562   

       cap-surface_g cap-surface_s cap-surface_y cap-color_c cap-color_e  ...  \
count           6499          6499          6499        6499        6499  ...   
unique             2             2             2           2           2  ...   
top            False         False         False       False       False  ...   
freq            6496          4449          3929        6464        5319  ...   

       population_n population_s population_v population_y habitat_g  \
count          6499         6499         6499         6499      6499   
unique            2            2            2            2         2   
top           False        False        False        False     False   
freq           6165         5509         3281         5113      4770   

       habitat_l habitat_m habitat_p habitat_u habitat_w  
count       6499      6499      6499      6499      6499  
unique         2         2         2         2         2  
top        False     False     False     False     False  
freq        5839      6268      5608      6198      6342  

[4 rows x 91 columns]

In [109]:
# Training set infomation on y_train
y_train.info()

<class 'pandas.core.series.Series'>
Index: 6499 entries, 5249 to 2411
Series name: poisonous_p
Non-Null Count  Dtype
--------------  -----
6499 non-null   bool 
dtypes: bool(1)
memory usage: 57.1 KB


In [110]:
# Training set description of X_train
y_train.describe()

count      6499
unique        2
top       False
freq       3366
Name: poisonous_p, dtype: object